src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp

```diff
@@ -0,0 +1,235 @@
+#include <Common/getRandomASCIIString.h>
+#include <Common/CurrentThread.h>
+#include <Common/MemoryTrackerBlockerInThread.h>
+#include <Common/logger_useful.h>
+#include <Common/scope_guard_safe.h>
+#include <Common/setThreadName.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ROCKSDB_ERROR;
+}
+
+static const IColumn::Permutation & getAscendingPermutation(const IColumn & column, IColumn::Permutation & perm)
+{
+    column.getPermutation(IColumn::PermutationSortDirection::Ascending, IColumn::PermutationSortStability::Stable, 0, 1, perm);
+    return perm;
+}
+
+/// Build SST file from key-value pairs
+static rocksdb::Status buildSSTFile(const String & path, const ColumnString & keys, const ColumnString & values, const std::optional<IColumn::Permutation> & perm_ = {})
+{
+    /// rocksdb::SstFileWriter requires keys to be sorted in ascending order
+    IColumn::Permutation calculated_perm;
+    const IColumn::Permutation & perm = perm_ ? *perm_ : getAscendingPermutation(keys, calculated_perm);
+
+    rocksdb::SstFileWriter sst_file_writer(rocksdb::EnvOptions{}, rocksdb::Options{});
+    auto status = sst_file_writer.Open(path);
+    if (!status.ok())
+        return status;
+
+    auto rows = perm.size();
+    WriteBufferFromOwnString wb_value;
+    for (size_t i = 0; i < rows; ++i)
+    {
+        auto row = perm[i];
+
+        status = sst_file_writer.Put(keys.getDataAt(row).toView(), values.getDataAt(row).toView());
+
+        /// There could be duplicated keys in chunk, thus Put may give IsInvalidArgument. This is ok, as we're certain that
+        /// keys are sorted in ascending order.
+        if (!status.ok() && !status.IsInvalidArgument())
+            return status;
+    }
+
+    sst_file_writer.Finish();
+    return rocksdb::Status::OK();
+}
+
+EmbeddedRocksDBBulkSink::EmbeddedRocksDBBulkSink(
+    ContextPtr context_, StorageEmbeddedRocksDB & storage_, const StorageMetadataPtr & metadata_snapshot_)
+    : SinkToStorage(metadata_snapshot_->getSampleBlock()), WithContext(context_), storage(storage_), metadata_snapshot(metadata_snapshot_)
+{
+    for (const auto & elem : getHeader())
+    {
+        if (elem.name == storage.primary_key)
+            break;
+        ++primary_key_pos;
+    }
+
+    serializations = getHeader().getSerializations();
+    min_block_size_rows = std::max(storage.getSettings().bulk_insert_block_size, getContext()->getSettingsRef().min_insert_block_size_rows);
+
+    /// If max_insert_threads > 1 we may have multiple EmbeddedRocksDBBulkSink and getContext()->getCurrentQueryId() is not guarantee to
+    /// to have a distinct path. Also we cannot use query id as directory name here, because it could be defined by user and not suitable
+    /// for directory name
+    auto base_directory_name = sipHash128String(getContext()->getCurrentQueryId());
+    insert_directory_queue = fs::path(storage.getDataPaths()[0]) / (base_directory_name + "-" + getRandomASCIIString(8));
```

Comment:
Have you considered storing temporary files in [`tmp_path`](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#tmp-path)? As I see when we store it directly in storage path we can avoid copying. But storing in tmp also has some advantages, for example it can be cleaned up on startup in case of server failures.

Reply:
I still prefer to write temporary files in same disk. But I agree that we need to handle garbage directories when server failures. I think we can do this: temporary insert directory will have a prefix (for example, `tmp_bulk_sink_`), then in `StorageEmbeddedRocksDB::initDB()` we can scan temporary insert directory and remove them before initializing rocksdb database. @vdimir Wdyt about this approach?

Reply:
I still prefer to have temporary files in same disk. But I agree we need to handle garbage when server failures. I think we can do it by having a specific prefix for temporary insert directory and then clean it in `StorageEmbeddedRocksDB::initDB()`. Wdyt?